### PR TITLE
fix(teacher_tab): partly fix Open/Close handling w.r.t. Assignments

### LINKED ISSUE #534: Bug: the Closed/Open state should appear for Assigned exercises
### Related user(s):

@erikmd, @AltGr

### Related issue(s) or PR(s):

_No response_

### Related project scope(s):

teacher dashboard

### Bug description:

If an exercise has at least one assigned student, the teacher dashboard displays the `Assigned` state only, while it should display `Open Assigned` or `Closed Assigned`, to show what the exercise status is for unassigned students.

### To reproduce:

_No response_

### Expected behavior:

_No response_

### learn-ocaml --version

_No response_

### git describe --long --always --abbrev=40 --tags

_No response_

### What OS do you use?

GNU/Linux

### What OS version/distribution do you use?

Debian

### What browser(s) do you use with learn-ocaml?

Firefox, Chromium

### What browser(s) version did you used to reproduce the issue?

_No response_

### Screenshots (if need be):

_No response_

### Additional context:

_No response_

### LINKED ISSUE #558: Bug: the Closed/Open state is buggy/confusing
### Related user(s):

@erikmd, @AltGr

### Related issue(s) or PR(s):

- Issue #534
- PR #549

### Related project scope(s):

teacher dashboard

### Bug description:

cf. [sync.zip](https://github.com/ocaml-sf/learn-ocaml/files/12326975/sync.zip)

With this zipped sync folder, there are two issues in the teacher dashboard:

1. if we click on the first exercise (demo) then on Open/Close once or several times, it's an **unexpected no-op** (the badge stays "Assigned", and after clicking on Apply, the sync/exercises.json file does not change either)
2. the second exercise (demo2) is **wrongly labelled "Assigned"**, while the exercise is not included in any assignment anymore.

FTR, here is the contents (pretty-printed with jq) of the file sync/exercises.json:

```json
[
  {
    "learnocaml_version": "2",
    "id": "demo",
    "assignments": {
      "default": {
        "start": 1691877600,
        "stop": 1692482400
      },
      "token_map": {
        "KOK-W3L-NE1-SVY": "Open"
      }
    }
  },
  {
    "learnocaml_version": "2",
    "id": "demo2",
    "assignments": {
      "default": "Closed",
      "token_map": {
        "KOK-W3L-NE1-SVY": "Open"
      }
    }
  }
]
```

### To reproduce:

Steps to reproduce the attached [sync.zip](https://github.com/ocaml-sf/learn-ocaml/files/12326975/sync.zip)

1. Create an empty sync folder
2. Run `learn-ocaml build serve --repo=demo-repository` from master branch
3. Create a user "Bar"
4. Create a user "Baz"
5. Open the teacher dashboard
6. Add an assignment including both exercises (demo, demo2) and only users (Bar, `<any future students>`)
7. Click on Apply
8. Click on the assignment, remove exercise demo2
9. Click on Apply

### Expected behavior:

_No response_

### learn-ocaml --version

_No response_

### git describe --long --always --abbrev=40 --tags

v0.14.1-39-g48583ba4d1376ad006227d7b5ee1b9cfaf9ad8f7

### What OS do you use?

GNU/Linux

### What OS version/distribution do you use?

Debian

### What browser(s) do you use with learn-ocaml?

Firefox, Chromium

### What browser(s) version did you used to reproduce the issue?

Firefox

### Screenshots (if need be):

_No response_

### Additional context:

_No response_

### PR DESCRIPTION
**Pull request by erikmd**
* **Kind:** bugfix
* Close #534
* Related: #558

### Description

* Refactor "Open"/"OpenAssg"… display code, Add functions in `Learnocaml_data` rather than in the js_of_ocaml code
* Make invariants explicit in OCaml comments
* Add unit tests using `ppx_expect`, `ppx_inline_test`

### Checklist

* [x] Read the [CONTRIBUTING.md](https://github.com/ocaml-sf/learn-ocaml/blob/master/CONTRIBUTING.md) guide and:
  * [x] Use [Atomic Commits](https://github.com/ocaml-sf/learn-ocaml/blob/master/CONTRIBUTING.md#atomic-commits) so each commit gathers a single logical change
  * [x] Use [Conventional Commits](https://github.com/ocaml-sf/learn-ocaml/blob/master/CONTRIBUTING.md#conventional-commits) regarding commit messages (needed by our release toolchain)
* [x] Add/update [tests](https://github.com/ocaml-sf/learn-ocaml/tree/master/tests#readme)
  
* [ ] Add/update [documentation](https://github.com/ocaml-sf/learn-ocaml/tree/master/docs)
  
* [ ] …
  

### Note to maintainers

* Read [this wiki page](https://github.com/ocaml-sf/learn-ocaml/wiki/Checklist-for-testing-and-merging-a-PR).
* Make sure the PR has a **milestone**.
* ***Assign*** yourself before merging.
* [x] Either do a regular **merge**:
  * for PRs containing *several commits* following [conventional-commits](https://github.com/ocaml-sf/learn-ocaml/blob/master/CONTRIBUTING.md#conventional-commits),
  * or for PRs containing 1 commit shared with a later PR (to preserve the SHA1)
* [ ] Or do a **squash-merge**:
  * for PRs containing *only 1 commit* (not shared with a later PR),
  * or for PRs containing several commits that need not be kept in the history;
  * → ***Update the commit message header*** with a [conventional-commit type](https://github.com/ocaml-sf/learn-ocaml/blob/master/CONTRIBUTING.md#conventional-commits-examples),
  * → ***Add a footer*** `Close #…` if a related issue exists.

### CODE REVIEW

File: src/app/learnocaml_teacher_tab.ml [hunk 1]
```diff
@@ -299,13 +299,11 @@ let rec teacher_tab token _select _params () =
...
-                  | true, ES.Open -> "exo_open", [%i"Open"]
-                  | true, ES.Closed -> "exo_closed", [%i"Closed"]
-                  | _, (ES.Assigned _ | ES.Closed) ->
-                      "exo_assigned", [%i"Assigned"]
-                  | false, ES.Open -> "exo_assigned", [%i"Open/Assg"]
+                  match ES.(is_open_assg_globally st.assignments) with
+                  | `Open -> "exo_open", [%i"Open"]
+                  | `OpenAssg -> "exo_assigned", [%i"Open/Assg"]
+                  | `Assigned -> "exo_assigned", [%i"Assigned"]
+                  | `Closed -> "exo_closed", [%i"Closed"]
```
Comment (yurug) [inline src/app/learnocaml_teacher_tab.ml:306]: Nitpicking: I am not a big fan of writing "Assg" for "Assignment"/"Assigned".

File: src/app/learnocaml_teacher_tab.ml [hunk 2]
```diff
@@ -299,13 +299,11 @@ let rec teacher_tab token _select _params () =
...
                 let cls, text =
-                  match Token.Map.is_empty ES.(st.assignments.token_map),
-                        ES.(st.assignments.default) with
-                  | true, ES.Open -> "exo_open", [%i"Open"]
-                  | true, ES.Closed -> "exo_closed", [%i"Closed"]
-                  | _, (ES.Assigned _ | ES.Closed) ->
-                      "exo_assigned", [%i"Assigned"]
-                  | false, ES.Open -> "exo_assigned", [%i"Open/Assg"]
+                  match ES.(is_open_assg_globally st.assignments) with
+                  | `Open -> "exo_open", [%i"Open"]
```
Comment (yurug) [inline src/app/learnocaml_teacher_tab.ml:303]: Why did you turn the result type into a polymorphic variant? (Polymorphic variants make sense for extensible code but they provide less guarantee to check for exhaustiveness...)

File: src/state/learnocaml_data.ml [hunk 3]
```diff
@@ -513,6 +513,57 @@ module Exercise = struct
       | Some a -> a
       | None -> a.default
 
+  (* Open/Close invariants:
+     forall exo_status,
+     1.(REQUIRED):
+        (exo_status.assignments.default <> Open /\ Token.Map.for_all (fun _ st -> st <> Open))
```
Comment (yurug) [inline src/state/learnocaml_data.ml:519]: I have some troubles understanding this formula because of the number of cases. For instance, it looks like you can have `exo_status.assignments.default = Close` and some token status equals to `Assigned x` for some `x`. Is that correct?

What is the type of `st` here?

Do you really want that, or:
```
forall exo_status e, exists x, e.assignments.default <> Assigned x -> forall known token t, t = e.assignments.default?
```

File: src/state/learnocaml_data.mli [hunk 4]
```diff
@@ -206,6 +206,13 @@ module Exercise: sig
     val get_status:
       Token.t -> assignments -> status
 
+    val is_open_assg_globally:
+      assignments -> [> `Open | `OpenAssg | `Assigned | `Closed]
+
+    val set_close_assg_globally: assignments -> assignments
+
+    val set_open_assg_globally: assignments -> assignments
```
Comment (yurug) [inline src/state/learnocaml_data.mli:214]: I feel the need for a docstring here, as the semantics of "OpenAssg" is unclear to me.

File: src/app/learnocaml_teacher_tab.ml (hunk 2 above)
Comment (erikmd) [reply src/app/learnocaml_teacher_tab.ml:303]: > Why did you turn the result type into a polymorphic variant? (Polymorphic variants make sense for extensible code but they provide less guarantee to check for exhaustiveness...)

Good point, it may be better to use a regular sum type; unfortunately we should use other constructors than `Open`, `Closed` to avoid any clash; maybe `GloballyOpen, GloballyClosed, GloballyOpenOrAssigned, GloballyClosedOrAssigned`

File: src/app/learnocaml_teacher_tab.ml (hunk 1 above)
Comment (erikmd) [reply src/app/learnocaml_teacher_tab.ml:306]: Maybe `Open/Assigned`

File: src/state/learnocaml_data.ml (hunk 3 above)
Comment (erikmd) [reply src/state/learnocaml_data.ml:519]: As we said in the telco, this is mostly a (map + default value) to emulate a constant function except for specific tokens.

We just don't want to have at the same time Open and Closed states, so that there is only 4 possible global states.

File: src/state/learnocaml_data.mli (hunk 4 above)
Comment (erikmd) [reply src/state/learnocaml_data.mli:214]: Good point.

**Comment by erikmd** [comment]
Summary of the telco with @yurug → ideally, we'd want:

1. a bit more doc (ocaml docstrings)
2. a check predicate to check an `assignments` value
3. a check_and_fix function to replace all spurious Open with Close 
4. a runtime test (learn-ocaml serve) that calls the check predicate and fails noisily if need be 
5. a CLI option that calls the check_and_fix function
6. unit tests that exercise these functions in a few corner cases

**Comment by erikmd** [comment]
@AltGr @yurug the PR is ready! except for the static builds failure (cf. this [discuss thread](https://discuss.ocaml.org/t/how-to-keep-static-builds-as-is-while-adding-ppx-expect-support-for-dev-test-builds/12915))

FYI, to follow-up the [previous comment](https://github.com/ocaml-sf/learn-ocaml/pull/559#issuecomment-1683885630):

On second thought, I didn't add a blocking runtime test (which could be distracting for end users), and improved the teacher tab instead to ensure the Inconsistency is displayed, and fixed on the next click on `Open/Close`.

See the following screenshots.

## Opening the teacher tab after loading the inconsistent [sync.zip](https://github.com/ocaml-sf/learn-ocaml/files/12326975/sync.zip) from #558

2023-08-28_14-23-45_Screenshot_sync zip

## After clicking on Open/Close for all exercises

2023-08-28_14-24-37_Screenshot_openclose

**Comment by erikmd** [comment]
Good news, the CI issue is now fixed.

Note: there's still an issue in learnocaml_teacher_tab.ml that may create Inconsistent exercise statuses (cf. #558)

Will fix this ASAP

**Review by AltGr — APPROVED** [review]
LGTM!

**Comment by yurug** [comment]
LGTM

**Review by yurug — APPROVED** [review]
